### PR TITLE
Fix: Correct toggling and feedback for grouped sources, nested scenes, sources in multiple scenes

### PR DESCRIPTION
**Pull request by bryce-seifert**
Fixes https://github.com/bitfocus/companion-module-obs-studio/issues/65
Fixes https://github.com/bitfocus/companion-module-obs-studio/issues/78
Fixes https://github.com/bitfocus/companion-module-obs-studio/issues/79

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -195,11 +194,11 @@ instance.prototype.init = function() {
 		});
 		
 		self.obs.on('SceneItemVisibilityChanged', function(data) {
-			if (self.states['studio_mode'] == true) { 
-				//Item in preview, no change
-			} else {
 				self.updateScenesAndSources();
```
Comment (astellingwerf) [inline index.js:197]: Indentation seems off.

File: index.js [hunk 2]
```diff
@@ -195,11 +194,11 @@ instance.prototype.init = function() {
...
-			}
+		});
+
+		self.obs.on('SceneItemTransformChanged', function(data) {
+				self.updateScenesAndSources();
```
Comment (astellingwerf) [inline index.js:201]: Indentation seems off.

File: index.js [hunk 3]
```diff
@@ -430,6 +427,82 @@ instance.prototype.updateScenesAndSources = async function() {
 		}
 	}
 
+	let updateSceneSources = (source, scene) => {
```
Comment (astellingwerf) [inline index.js:430]: Wouldn't this function be easier to read of you'd store `self.sources[source.name]` in a `const`?

File: index.js [hunk 4]
```diff
@@ -948,26 +1020,37 @@ instance.prototype.action = function(action) {
...
 					let scene = self.scenes[sceneName]
 					if (scene) {
 						for (let source of scene.sources) {
-							if (source.name == action.options.source) {
+							if (source.type == 'scene' && source.name == action.options.source) {
+								visible = !source.render
+							}
+							else if (source.type == 'group') {
+								if (source.name == action.options.source) {
+								visible = !source.render
```
Comment (astellingwerf) [inline index.js:1034]: Indentation seems off.

File: index.js [hunk 5]
```diff
@@ -430,6 +427,82 @@ instance.prototype.updateScenesAndSources = async function() {
...
+				}
+			}
+		}
+	}
+
+	let updateRegularSources = (source, scene) => {
+		if (self.sources[source.name] && self.sources[source.name]['visible'] === true) {
+			self.sources[source.name]['visible'] = true;
+		} else {
+			if (source.render === true && scene.name && scene.name == sceneList.currentScene) {
```
Comment (astellingwerf) [inline index.js:486]: Could be merged into `else if`.

File: index.js [hunk 6]
```diff
@@ -430,6 +427,82 @@ instance.prototype.updateScenesAndSources = async function() {
...
+		for (let source of scene.sources) {
+			if (source.type == 'scene') {
+				updateSceneSources(source, scene)
+			}
+			if (source.type == 'group') {
+				updateGroupedSources(source, scene)
+			}
+			if (source.type !== 'group' && source.type !== 'scene') {
+				updateRegularSources(source, scene)
+			}
```
Comment (astellingwerf) [inline index.js:502]: I'd write this as:

```suggestion
			if (source.type == 'scene') {
				updateSceneSources(source, scene)
			} else if (source.type == 'group') {
				updateGroupedSources(source, scene)
			} else {
				updateRegularSources(source, scene)
			}
```

File: index.js [hunk 7]
```diff
@@ -430,6 +427,82 @@ instance.prototype.updateScenesAndSources = async function() {
...
+		if (source.render === true && self.sources[source.name] && scene.name == sceneList.currentScene) {
+			self.sources[source.name]['visible'] = true;
+		}
+		if (source.render === true) {
+			for (var s in source.groupChildren) {
+				var groupedSource = source.groupChildren[s];
+				if (groupedSource.type == 'scene') {
+					updateSceneSources(groupedSource, scene)
+				} else {
+					if (groupedSource.render === true && source.render === true && scene.name == sceneList.currentScene) {
```
Comment (astellingwerf) [inline index.js:474]: Could be merged into `else if`.

File: index.js [hunk 8]
```diff
@@ -430,6 +427,82 @@ instance.prototype.updateScenesAndSources = async function() {
...
+			}
+		}
+	}
+
+	let updateGroupedSources = (source, scene) => {
+		if (source.render === true && self.sources[source.name] && scene.name == sceneList.currentScene) {
+			self.sources[source.name]['visible'] = true;
+		}
+		if (source.render === true) {
+			for (var s in source.groupChildren) {
```
Comment (astellingwerf) [inline index.js:469]: You're mixing `for (var s in ...)` with  `for (let s in ...)` throughout this PR.

File: index.js (hunk 1 above)
Comment (bryce-seifert) [reply index.js]: Fixed

File: index.js (hunk 2 above)
Comment (bryce-seifert) [reply index.js]: Fixed

File: index.js (hunk 4 above)
Comment (bryce-seifert) [reply index.js]: Fixed

File: index.js (hunk 5 above)
Comment (bryce-seifert) [reply index.js]: Fixed

File: index.js (hunk 6 above)
Comment (bryce-seifert) [reply index.js]: Done

File: index.js (hunk 7 above)
Comment (bryce-seifert) [reply index.js]: Done

File: index.js (hunk 8 above)
Comment (bryce-seifert) [reply index.js]: Changed this to be consistent for all

File: index.js (hunk 3 above)
Comment (bryce-seifert) [reply index.js:430]: I'm not quite sure how to do this. I tried declaring this within each block, but it seems to keep ending up in a TypeError with the const being undefined.

Comment (bryce-seifert) [reply index.js:430]: I think the problem I'm running into if I try and do this is that sources are not always defined yet when running through this process, so some aren't assigned to self.sources until later in the process.

**Pull request merged**